File: go/tpcc_benchmark.go
package main

import (
	"context"
	"fmt"
	"log"
	"math/rand"
	"os"
	"os/signal"
	"sync"
	"syscall"
	"time"

	"cloud.google.com/go/spanner"
	"github.com/urfave/cli/v3"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"
)

func executeTPCCBenchmark(ctx context.Context, cmd *cli.Command) error {
	runCtx, cancel := context.WithCancel(ctx)
	defer cancel()

	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigs
		cancel()
	}()

	cfg := parseGlobalConfig(cmd)
	warehouses := int(cmd.Int("warehouses"))
	clients := int(cmd.Int("clients"))
	items := int(cmd.Int("items"))
	extended := cmd.Bool("extended")

	latencyHistogram, _, operationCounter, errorCounter, memoryUsageHistogram, cpuUtilizationHistogram, cleanupMetrics, err := setupMetrics(runCtx, cfg.Project, cfg.Host, cfg.BenchmarkName)
	if err != nil {
		return fmt.Errorf("failed to initialize metrics: %w", err)
	}
	defer cleanupMetrics()

	client, err := createSpannerClient(runCtx, cfg.Project, cfg.Instance, cfg.Database, cfg.Host)
	if err != nil {
		return fmt.Errorf("failed to build Spanner client: %w", err)
	}
	defer client.Close()

	baseAttributeList := []attribute.KeyValue{
		attribute.String("benchmark_type", "tpcc"),
		attribute.Bool("for_alerting", cfg.ForAlerting),
		attribute.String("benchmark_name", cfg.BenchmarkName),
		attribute.String("client", "go-client"),
		attribute.Int("concurrent_clients", clients),
	}
	if extended {
		baseAttributeList = append(baseAttributeList, attribute.Bool("extended", true))
	}
	attributes := metric.WithAttributes(baseAttributeList...)

	attrNewOrder := metric.WithAttributes(append(baseAttributeList, attribute.String("transaction_type", "new_order"))...)
	attrNewOrderMutations := metric.WithAttributes(append(baseAttributeList, attribute.String("transaction_type", "new_order_mutations"))...)
	attrPayment := metric.WithAttributes(append(baseAttributeList, attribute.String("transaction_type", "payment"))...)
	attrPaymentMutationsDirect := metric.WithAttributes(append(baseAttributeList, attribute.String("transaction_type", "payment_mutations_direct"))...)
	attrOrderStatus := metric.WithAttributes(append(baseAttributeList, attribute.String("transaction_type", "order_status"))...)
	attrOrderStatusReads := metric.WithAttributes(append(baseAttributeList, attribute.String("transaction_type", "order_status_reads"))...)
	attrDelivery := metric.WithAttributes(append(baseAttributeList, attribute.String("transaction_type", "delivery"))...)
	attrStockLevel := metric.WithAttributes(append(baseAttributeList, attribute.String("transaction_type", "stock_level"))...)
	attrStockLevelPartitioned := metric.WithAttributes(append(baseAttributeList, attribute.String("transaction_type", "stock_level_partitioned"))...)

	// Setup duration context timeout if not infinite
	var durationCtx context.Context
	if cfg.DurationStr != "inf" && cfg.DurationStr != "infinite" {
		duration := ParseDuration(cfg.DurationStr)
		if duration > 0 {
			var durationCancel context.CancelFunc
			durationCtx, durationCancel = context.WithTimeout(runCtx, duration)
			defer durationCancel()
		} else {
			durationCtx = runCtx
		}
	} else {
		durationCtx = runCtx
	}

	startResourceMonitoring(durationCtx, memoryUsageHistogram, cpuUtilizationHistogram, cfg.ResourceProbeInterval, attributes)

	extendedModeStr := ""
	if extended {
		extendedModeStr = " [EXTENDED MODE]"
	}
	fmt.Printf("Starting TPC-C Benchmark for %s, Scale Factor (Warehouses): %d, Parallel Clients: %d, Items: %d%s\n", cfg.DurationStr, warehouses, clients, items, extendedModeStr)

	// Assert database capacity
	singleIter := client.Single().Query(durationCtx, spanner.Statement{SQL: "SELECT COUNT(*) FROM warehouse"})
	row, err := singleIter.Next()
	singleIter.Stop()
	if err != nil {
		return fmt.Errorf("failed to query warehouse count: %w", err)
	}
	if row != nil {
		var warehouseCount int64
		if err := row.Column(0, &warehouseCount); err != nil {
			return fmt.Errorf("failed to parse warehouse count: %w", err)
		}
		if warehouseCount < int64(warehouses) {
			return fmt.Errorf("database capacity check failed: Required scale factor %d warehouses, but database only has %d", warehouses, warehouseCount)
		}
	}

	wg := &sync.WaitGroup{}

	for i := 0; i < clients; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for {
				select {
				case <-durationCtx.Done():
					return
				default:
					prob := rand.Intn(100)
					var err error
					var attr metric.MeasurementOption
					txType := "new_order"

					start := time.Now()
					if extended {
						if prob < 25 {
							txType = "new_order"
							attr = attrNewOrder
							err = executeNewOrder(durationCtx, client, warehouses, items, true)
						} else if prob < 45 {
							txType = "new_order_mutations"
							attr = attrNewOrderMutations
							err = executeNewOrderMutations(durationCtx, client, warehouses, items)
						} else if prob < 78 {
							txType = "payment"
							attr = attrPayment
							err = executePayment(durationCtx, client, warehouses, true)
						} else if prob < 88 {
							txType = "payment_mutations_direct"
							attr = attrPaymentMutationsDirect
							err = executePaymentMutationsDirect(durationCtx, client, warehouses)
						} else if prob < 90 {
							txType = "order_status"
							attr = attrOrderStatus
							err = executeOrderStatus(durationCtx, client, warehouses, true)
						} else if prob < 92 {
							txType = "order_status_reads"
							attr = attrOrderStatusReads
							err = executeOrderStatusReads(durationCtx, client, warehouses)
						} else if prob < 96 {
							txType = "delivery"
							attr = attrDelivery
							err = executeDelivery(durationCtx, client, warehouses, true)
						} else if prob < 98 {
							txType = "stock_level"
							attr = attrStockLevel
							err = executeStockLevel(durationCtx, client, warehouses, true)
						} else {
							txType = "stock_level_partitioned"
							attr = attrStockLevelPartitioned
							err = executeStockLevelPartitioned(durationCtx, client, warehouses)
						}
					} else {
						if prob < 45 {
							txType = "new_order"
							attr = attrNewOrder
							err = executeNewOrder(durationCtx, client, warehouses, items, false)
						} else if prob < 88 {
							txType = "payment"
							attr = attrPayment
							err = executePayment(durationCtx, client, warehouses, false)
						} else if prob < 92 {
							txType = "order_status"
							attr = attrOrderStatus
							err = executeOrderStatus(durationCtx, client, warehouses, false)
						} else if prob < 96 {
							txType = "delivery"
							attr = attrDelivery
							err = executeDelivery(durationCtx, client, warehouses, false)
						} else {
							txType = "stock_level"
							attr = attrStockLevel
							err = executeStockLevel(durationCtx, client, warehouses, false)
						}
					}

					if err != nil {
						if durationCtx.Err() == nil {
							log.Printf("TPC-C transaction %s failed: %v", txType, err)
							errorCounter.Add(runCtx, 1, attr)
						}
					} else {
						latencyHistogram.Record(runCtx, float64(time.Since(start).Microseconds()), attr)
					}
					operationCounter.Add(runCtx, 1, attr)
				}
			}
		}()
	}

	wg.Wait()
	fmt.Println("TPC-C benchmark execution complete.")
	return nil
}

File: java/src/main/java/com/google/cloud/spanner/benchmark/tpcc/TpccBenchmark.java
package com.google.cloud.spanner.benchmark.tpcc;

import com.google.cloud.spanner.BatchClient;
import com.google.cloud.spanner.DatabaseClient;
import com.google.cloud.spanner.ResultSet;
import com.google.cloud.spanner.Statement;
import com.google.cloud.spanner.benchmark.AbstractBenchmark;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.common.AttributesBuilder;
import io.opentelemetry.api.metrics.DoubleHistogram;
import io.opentelemetry.api.metrics.LongCounter;
import io.opentelemetry.api.metrics.LongHistogram;
import java.time.Duration;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

public class TpccBenchmark {

  private final DatabaseClient client;
  private final BatchClient batchClient;
  private final LongHistogram latencyHistogram;
  private final LongCounter operationCounter;
  private final LongCounter errorCounter;
  private final int scaleFactor;
  private final int clients;
  private final int items;
  private final Duration duration;
  private final boolean extended;
  private final Attributes baseAttributes;
  private final Attributes newOrderAttributes;
  private final Attributes newOrderMutationsAttributes;
  private final Attributes paymentAttributes;
  private final Attributes paymentMutationsDirectAttributes;
  private final Attributes orderStatusAttributes;
  private final Attributes orderStatusReadsAttributes;
  private final Attributes deliveryAttributes;
  private final Attributes stockLevelAttributes;
  private final Attributes stockLevelPartitionedAttributes;

  private final LongHistogram memoryUsageHistogram;
  private final DoubleHistogram cpuUtilizationHistogram;
  private final String resourceProbeInterval;
  private java.util.concurrent.ScheduledExecutorService resourceMonitorExecutor;

  public TpccBenchmark(
      DatabaseClient client,
      BatchClient batchClient,
      LongHistogram latencyHistogram,
      LongCounter operationCounter,
      LongCounter errorCounter,
      LongHistogram memoryUsageHistogram,
      DoubleHistogram cpuUtilizationHistogram,
      String resourceProbeInterval,
      int scaleFactor,
      int clients,
      int items,
      Duration duration,
      boolean forAlerting,
      String benchmarkName,
      boolean extended) {
    this.client = client;
    this.batchClient = batchClient;
    this.latencyHistogram = latencyHistogram;
    this.operationCounter = operationCounter;
    this.errorCounter = errorCounter;
    this.memoryUsageHistogram = memoryUsageHistogram;
    this.cpuUtilizationHistogram = cpuUtilizationHistogram;
    this.resourceProbeInterval = resourceProbeInterval;
    this.scaleFactor = scaleFactor;
    this.clients = clients;
    this.items = items;
    this.duration = duration;
    this.extended = extended;

    AttributesBuilder baseAttributesBuilder =
        Attributes.builder()
            .put("benchmark_type", "tpcc")
            .put("for_alerting", forAlerting)
            .put("benchmark_name", benchmarkName != null ? benchmarkName : "")
            .put("client", "java-client")
            .put("concurrent_clients", clients);
    if (extended) {
      baseAttributesBuilder.put("extended", true);
    }
    this.baseAttributes = baseAttributesBuilder.build();

    this.newOrderAttributes =
        baseAttributes.toBuilder().put("transaction_type", "new_order").build();
    this.newOrderMutationsAttributes =
        baseAttributes.toBuilder().put("transaction_type", "new_order_mutations").build();
    this.paymentAttributes = baseAttributes.toBuilder().put("transaction_type", "payment").build();
    this.paymentMutationsDirectAttributes =
        baseAttributes.toBuilder().put("transaction_type", "payment_mutations_direct").build();
    this.orderStatusAttributes =
        baseAttributes.toBuilder().put("transaction_type", "order_status").build();
    this.orderStatusReadsAttributes =
        baseAttributes.toBuilder().put("transaction_type", "order_status_reads").build();
    this.deliveryAttributes =
        baseAttributes.toBuilder().put("transaction_type", "delivery").build();
    this.stockLevelAttributes =
        baseAttributes.toBuilder().put("transaction_type", "stock_level").build();
    this.stockLevelPartitionedAttributes =
        baseAttributes.toBuilder().put("transaction_type", "stock_level_partitioned").build();
  }

  public void run() {
    System.out.println(
        "Starting TPC-C Benchmark with Scale Factor (Warehouses): "
            + scaleFactor
            + ", Parallel Clients: "
            + clients
            + ", Items: "
            + items
            + (extended ? " [EXTENDED MODE]" : ""));

    startResourceMonitoring();

    // Assert database capacity
    try (ResultSet rs =
        client.singleUse().executeQuery(Statement.of("SELECT COUNT(*) FROM warehouse"))) {
      if (rs.next()) {
        long warehouseCount = rs.getLong(0);
        if (warehouseCount < scaleFactor) {
          throw new IllegalStateException(
              "Database capacity check failed: Required scale factor "
                  + scaleFactor
                  + " warehouses, but database only has "
                  + warehouseCount);
        }
      }
    }

    ExecutorService executor = Executors.newFixedThreadPool(clients);
    long startTimeMs = System.currentTimeMillis();

    for (int c = 0; c < clients; c++) {
      executor.submit(
          () -> {
            while (!Thread.currentThread().isInterrupted()) {
              if (duration != null
                  && (System.currentTimeMillis() - startTimeMs) >= duration.toMillis()) {
                break;
              }

              int prob = ThreadLocalRandom.current().nextInt(100);
              Attributes currentAttributes = baseAttributes;
              long startNs = System.nanoTime();
              boolean success = false;

              try {
                if (extended) {
                  if (prob < 25) {
                    currentAttributes = newOrderAttributes;
                    TpccTransactions.executeNewOrder(client, scaleFactor, items, true);
                  } else if (prob < 45) {
                    currentAttributes = newOrderMutationsAttributes;
                    TpccTransactions.executeNewOrderMutations(client, scaleFactor, items);
                  } else if (prob < 78) {
                    currentAttributes = paymentAttributes;
                    TpccTransactions.executePayment(client, scaleFactor, true);
                  } else if (prob < 88) {
                    currentAttributes = paymentMutationsDirectAttributes;
                    TpccTransactions.executePaymentMutationsDirect(client, scaleFactor);
                  } else if (prob < 90) {
                    currentAttributes = orderStatusAttributes;
                    TpccTransactions.executeOrderStatus(client, scaleFactor, true);
                  } else if (prob < 92) {
                    currentAttributes = orderStatusReadsAttributes;
                    TpccTransactions.executeOrderStatusReads(client, scaleFactor);
                  } else if (prob < 96) {
                    currentAttributes = deliveryAttributes;
                    TpccTransactions.executeDelivery(client, scaleFactor, true);
                  } else if (prob < 98) {
                    currentAttributes = stockLevelAttributes;
                    TpccTransactions.executeStockLevel(client, scaleFactor, true);
                  } else {
                    currentAttributes = stockLevelPartitionedAttributes;
                    TpccTransactions.executeStockLevelPartitioned(batchClient, scaleFactor);
                  }
                } else {
                  if (prob < 45) {
                    currentAttributes = newOrderAttributes;
                    TpccTransactions.executeNewOrder(client, scaleFactor, items, false);
                  } else if (prob < 88) {
                    currentAttributes = paymentAttributes;
                    TpccTransactions.executePayment(client, scaleFactor, false);
                  } else if (prob < 92) {
                    currentAttributes = orderStatusAttributes;
                    TpccTransactions.executeOrderStatus(client, scaleFactor, false);
                  } else if (prob < 96) {
                    currentAttributes = deliveryAttributes;
                    TpccTransactions.executeDelivery(client, scaleFactor, false);
                  } else {
                    currentAttributes = stockLevelAttributes;
                    TpccTransactions.executeStockLevel(client, scaleFactor, false);
                  }
                }
                success = true;
              } catch (Exception e) {
                e.printStackTrace();
                errorCounter.add(1, currentAttributes);
              } finally {
                if (success) {
                  long latencyUs = (System.nanoTime() - startNs) / 1000;
                  latencyHistogram.record(latencyUs, currentAttributes);
                }
                operationCounter.add(1, currentAttributes);
              }
            }
          });
    }

    try {
      if (duration != null) {
        Thread.sleep(duration.toMillis());
      } else {
        Thread.sleep(Long.MAX_VALUE);
      }
      System.out.println("TPC-C duration complete. Shutting down pool...");
      if (resourceMonitorExecutor != null) {
        resourceMonitorExecutor.shutdownNow();
      }
      executor.shutdownNow();
      executor.awaitTermination(1, TimeUnit.MINUTES);
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      if (resourceMonitorExecutor != null) {
        resourceMonitorExecutor.shutdownNow();
      }
      executor.shutdownNow();
    }
  }

  private void startResourceMonitoring() {
    if (resourceProbeInterval != null && !resourceProbeInterval.isEmpty()) {
      Duration probeDuration = AbstractBenchmark.parseDuration(resourceProbeInterval);
      if (probeDuration != null && probeDuration.toMillis() > 0) {
        resourceMonitorExecutor = java.util.concurrent.Executors.newSingleThreadScheduledExecutor();
        resourceMonitorExecutor.scheduleAtFixedRate(
            this::probeResourceUsage, 0, probeDuration.toMillis(), TimeUnit.MILLISECONDS);
      }
    }
  }

  private void probeResourceUsage() {
    try {
      long usedMemory =
          java.lang.management.ManagementFactory.getMemoryMXBean().getHeapMemoryUsage().getUsed();
      if (memoryUsageHistogram != null) {
        memoryUsageHistogram.record(usedMemory, baseAttributes);
      }
      java.lang.management.OperatingSystemMXBean osBean =
          java.lang.management.ManagementFactory.getOperatingSystemMXBean();
      if (osBean instanceof com.sun.management.OperatingSystemMXBean) {
        double cpuLoad = ((com.sun.management.OperatingSystemMXBean) osBean).getProcessCpuLoad();
        if (cpuLoad >= 0 && cpuUtilizationHistogram != null) {
          cpuUtilizationHistogram.record(cpuLoad, baseAttributes);
        }
      }
    } catch (Exception e) {
      // Ignore exceptions in resource monitoring
    }
  }
}
